### ISSUE TITLE
backend discovery by name

### ISSUE THREAD

**Issue by frispete** [jaraco/keyring#253]
Hi, what's the preferred method of backend discovery?

I want to supply a GUI selector for keyrings to my program, because I noticed, that due to testing and fixing some SecretService build issue here, the keyring in use moved from KWallet to Secretservice behind the scenes without any further notice. (Well, it may asked for credentials one additional time, but because I'm debugging all this, so it's gone unnoticed as well).

Here's a selection of what I found so far.

Using the object itself:
```
--> for kr in keyring.backend.get_all_keyring(): print(kr)
... 
<keyring.backends.kwallet.DBusKeyringKWallet4 object at 0x7f03eb60fa90>
<keyring.backends.SecretService.Keyring object at 0x7f03eb60f438>
<keyring.backends.fail.Keyring object at 0x7f03eb60f9b0>
```
Fiddling with object names is no fun, of course.

`__class__.__name__`:
```
--> for kr in keyring.backend.get_all_keyring(): print(kr.__class__.__name__)
... 
DBusKeyringKWallet4
Keyring
Keyring
```
my favorite, but not suitable for obvious reasons....

`__module__`:
```
--> for kr in keyring.backend.get_all_keyring(): print(kr.__module__)
... 
keyring.backends.kwallet
keyring.backends.SecretService
keyring.backends.fail
```
That would be nice, but fails to differenciate the kwallets..

Hm, boils down to to the object so far. Not that nice. 

Am I missing something obvious?

Since the kwallet thing isn't released, we could throw it into another module (kwallet4), and have `__module__` work well, but I haven't looked into keyrings.alt, if there are multiple backend classes in a single module, that need similar differentiation. 

I would love to prepare a related patch, if this approach is acceptable.

**Comment by frispete** [comment]
Nobody with an opinion on this one? Jason?

**Comment by jaraco** [comment]
I'd like to avoid relying strictly on the module name, which as you've indicated would have implications on where one should store a keyring class.

A couple of things come to mind. You could use the module name + class name, which would yield:

```
kwallet.Keyring
kwallet.DBusKeyringKwallet4
SecretService.Keyring
fail.Keyring
...
```

Better would be for each class to present its own user-friendly name (and maybe description).

If the base class had this class property:

```

@ClassProperty
@classmethod
def name(cls):
    parent, sep, mod_name = cls.__module__.rpartition('.')
    mod_name = mod_name.replace('_', ' ')
    return ' '.join([mod_name, cls.__name__])
```

Then you could do:

```
names = [be.name for be in keyring.backend.get_all_keyring()]
```

And that would result in something like:

```
names == ['OS X Keyring', 'SecretService Keyring', 'kwallet Keyring', 'kwallet DBusKeyringKWallet4', ...]
```

And if any particular class doesn't present a suitable name, it can be overridden by setting `name` on the class.

How does that strike you?

**Comment by frispete** [comment]
It would strike me as "Wonderful!", since this is exactly, what a UI needs.

Should I go for a patch, or are you willing to commit this thing?

**Comment by jaraco** [comment]
Please do put together a PR. Thanks!

**Comment by jaraco** [comment]
Thanks for the contrib. Released in 10.1.
